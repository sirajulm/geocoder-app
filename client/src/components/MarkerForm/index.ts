import MarkerForm from './MarkerForm'

export default MarkerForm