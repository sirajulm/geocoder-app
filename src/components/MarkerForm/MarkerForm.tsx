import React, { Component, ChangeEvent, SyntheticEvent } from 'react';
import AutoSuggest from 'react-autosuggest';
import styled from 'styled-components';
import Input from '../Input';
import Button from '../Button';
import _ from 'lodash';

const StyledForm = styled.div`
    width: 70%
    margin: auto;
`;
const StyledFormTitle = styled.div`
    margin: 10px 15px 25px 15px;
    font-family: 'Open Sans', sans-serif;
    font-size: 24px;
    font-weight: 700;
`;

const StyledError = styled.div`
    font-family: 'Open Sans', sans-serif;
    color: red;
    margin-left: 15px;
`;

/**
 * Takes two object and returns the keys for which values have changed
 * @param o1 object
 * @param o2 object
 */
function changedKeys(o1: any, o2: any) {
    const keys = _.union(_.keys(o1), _.keys(o2));
    return _.filter(keys, (key) => {
        return o1[key] !== o2[key];
    })
}

interface MarkerFormProps {
    id?: string
    title: string;
    address: string;
    onSubmit: (query: any) => void;
};
interface MarkerFormState {
    title: string;
    address: string;
    error: string | null;

}

class MarkerForm extends Component<MarkerFormProps, MarkerFormState> {
    constructor(props: any) {
        super(props);
        this.state = {
            title: props.title || '',
            address: props.address || '',
            error: null
        }
    }

    onChange = (event: ChangeEvent<HTMLInputElement>) => {
        const newState = { [event.target.name]: event.target.value } as Pick<MarkerFormState, keyof MarkerFormState>
        this.setState(newState);
    };

    isValid = () => {
        const { title, address, error } = this.state;
        if (title.trim() === '' || address.trim() === '') {
            this.setState({ error: 'Some required fields are empty' })
            return false
        }
        this.setState({ error: null })
        return true
    }

    onSearch = (event: SyntheticEvent) => {
        const { title, address, error } = this.state;
        const { onSubmit } = this.props;
        if (this.isValid()) {
            const changedFields = _.pick(this.state, changedKeys({ title, address }, this.props))
            const updateData = { ...changedFields, id: this.props.id }
            onSubmit(updateData);
        }
    };

    render() {
        const { title, address, error } = this.state;

        return (
            <StyledForm>
                <StyledFormTitle>Add/Edit Marker</StyledFormTitle>
                {error &&
                    <StyledError>
                        {error}
                    </StyledError>
                }
                <Input
                    name="title"
                    value={title}
                    label="Title"
                    required={true}
                    placeholder="Enter Title"
                    onChange={this.onChange}
                />
                <Input
                    name="address"
                    value={address}
                    label="Address"
                    required={true}
                    placeholder="Enter Address"
                    onChange={this.onChange}
                />
                <Button
                    kind="primary"
                    onClick={this.onSearch}
                >
                    Save
                </Button>
            </StyledForm>
        );
    }
}

export default MarkerForm;