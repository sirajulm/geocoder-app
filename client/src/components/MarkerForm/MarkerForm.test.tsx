import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer'
import MarkerForm from './MarkerForm';


describe('MarkerForm', () => {
    it('should render', () => {
        const component = renderer.create(<MarkerForm />).toJSON()
        expect(component).toMatchSnapshot()
    })

    it('should render 2input fields with label and 1 button', () => {
        const component = mount(<MarkerForm />)
        expect(component.find('label')).toHaveLength(2)
        expect(component.find('input')).toHaveLength(2)
        expect(component.find('button')).toHaveLength(1)
    })

    describe('should trigger onSubmit of create', () => {
        const onSubmitMock = jest.fn();

        const component = mount(<MarkerForm onSubmit={onSubmitMock} />);
        it('should not trigger onSubmit with title and address empty', () => {
            const onChangeTitleEvent = {
                target: {
                    name: 'title',
                    value: ''
                }
            };

            const onChangeAddressEvent = {
                target: {
                    name: 'address',
                    value: ''
                }
            };
            component.find('input[name="title"]').simulate('change', onChangeTitleEvent);
            component.find('input[name="address"]').simulate('change', onChangeAddressEvent);

            expect(component.state().title).toBe(onChangeTitleEvent.target.value)
            expect(component.state().address).toBe(onChangeAddressEvent.target.value)
            const button = component.find('button').simulate('click');

            expect(onSubmitMock).not.toBeCalled()


        })

        it('should not trigger onSubmit with title alone', () => {
            const onChangeTitleEvent = {
                target: {
                    name: 'title',
                    value: 'sample title'
                }
            };

            const onChangeAddressEvent = {
                target: {
                    name: 'address',
                    value: ''
                }
            };
            component.find('input[name="title"]').simulate('change', onChangeTitleEvent);
            component.find('input[name="address"]').simulate('change', onChangeAddressEvent);

            expect(component.state().title).toBe(onChangeTitleEvent.target.value)
            expect(component.state().address).toBe(onChangeAddressEvent.target.value)
            const button = component.find('button').simulate('click');
            expect(onSubmitMock).not.toBeCalled()

        })
        it('should not trigger onSubmit with address alone', () => {
            const onChangeTitleEvent = {
                target: {
                    name: 'title',
                    value: ''
                }
            };

            const onChangeAddressEvent = {
                target: {
                    name: 'address',
                    value: 'sample address'
                }
            };
            component.find('input[name="title"]').simulate('change', onChangeTitleEvent);
            component.find('input[name="address"]').simulate('change', onChangeAddressEvent);

            expect(component.state().title).toBe(onChangeTitleEvent.target.value)
            expect(component.state().address).toBe(onChangeAddressEvent.target.value)
            const button = component.find('button').simulate('click');
            expect(onSubmitMock).not.toBeCalled()

        })

        it('should trigger onSubmit with title and address', () => {
            const onChangeTitleEvent = {
                target: {
                    name: 'title',
                    value: 'sample title'
                }
            };

            const onChangeAddressEvent = {
                target: {
                    name: 'address',
                    value: 'sample address'
                }
            };
            component.find('input[name="title"]').simulate('change', onChangeTitleEvent);
            component.find('input[name="address"]').simulate('change', onChangeAddressEvent);

            expect(component.state().title).toBe(onChangeTitleEvent.target.value)
            expect(component.state().address).toBe(onChangeAddressEvent.target.value)
            const button = component.find('button').simulate('click');
            expect(onSubmitMock).toBeCalled()
            expect(onSubmitMock).toBeCalledWith({ "address": "sample address", "id": undefined, "title": "sample title" })

        })
    })

    describe('should trigger onSubmit for update', () => {

        const title = 'sample title'
        const address = 'sample address'
        const id = 'sample-id';

        it('should trigger onSubmit with address when title not changed', () => {
            const onSubmitMock = jest.fn();
            const component = mount(<MarkerForm onSubmit={onSubmitMock} id={id} title={title} />);

            const onChangeAddressEvent = {
                target: {
                    name: 'address',
                    value: 'sample address'
                }
            };

            component.find('input[name="address"]').simulate('change', onChangeAddressEvent);

            expect(component.state().title).toBe(title)
            expect(component.state().address).toBe(onChangeAddressEvent.target.value)
            const button = component.find('button').simulate('click');
            expect(onSubmitMock).toBeCalledTimes(1)
            expect(onSubmitMock).toBeCalledWith({ address: 'sample address', id: 'sample-id' })

        })

        it('should trigger onSubmit with title when address not changed', () => {
            const onSubmitMock = jest.fn();
            const component = mount(<MarkerForm onSubmit={onSubmitMock} id={id} address={address} />);

            const onChangeTitleEvent = {
                target: {
                    value: 'sample title',
                    name: 'title',
                }
            };

            component.find('input[name="title"]').simulate('change', onChangeTitleEvent);

            expect(component.state().title).toBe(onChangeTitleEvent.target.value)
            expect(component.state().address).toBe(address)
            const button = component.find('button').simulate('click');
            expect(onSubmitMock).toBeCalled()
            expect(onSubmitMock).toBeCalledTimes(1)
            expect(onSubmitMock).toBeCalledWith({ "id": "sample-id", "title": "sample title" })

        })
    })
})
