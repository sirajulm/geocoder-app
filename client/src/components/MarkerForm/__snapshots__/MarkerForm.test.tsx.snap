// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerForm should render 1`] = `
<div
  className="sc-ifAKCX hFuvIb"
>
  <div
    className="sc-EHOje bVEYVe"
  >
    Add/Edit Marker
  </div>
  <div
    className="sc-bwzfXH ZRyws"
  >
    <label
      className="sc-bdVaJa cFGRTz"
      htmlFor="title"
    >
      Title
       *
    </label>
    <input
      className="sc-htpNat fDWxMd"
      name="title"
      onChange={[Function]}
      placeholder="Enter Title"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bwzfXH ZRyws"
  >
    <label
      className="sc-bdVaJa cFGRTz"
      htmlFor="address"
    >
      Address
       *
    </label>
    <input
      className="sc-htpNat fDWxMd"
      name="address"
      onChange={[Function]}
      placeholder="Enter Address"
      required={true}
      type="text"
      value=""
    />
  </div>
  <button
    className="sc-bxivhb kEfJWi"
    kind="primary"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
